pkg/indexer/atlasdatabaseuserprojects.go

```diff
@@ -0,0 +1,48 @@
+package indexer
+
+import (
+	"go.uber.org/zap"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	akov2 "github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api/v1"
+)
+
+const (
+	AtlasDatabaseUserByProjectsIndex = "atlasdatabaseuser.spec.project"
+)
+
+type AtlasDatabaseUserByProjectsIndexer struct {
+	logger *zap.SugaredLogger
+}
+
+func NewAtlasDatabaseUserByProjectsIndexer(logger *zap.Logger) *AtlasDatabaseUserByProjectsIndexer {
+	return &AtlasDatabaseUserByProjectsIndexer{
+		logger: logger.Named(AtlasDatabaseUserByProjectsIndex).Sugar(),
+	}
+}
+
+func (*AtlasDatabaseUserByProjectsIndexer) Object() client.Object {
+	return &akov2.AtlasDatabaseUser{}
+}
+
+func (*AtlasDatabaseUserByProjectsIndexer) Name() string {
+	return AtlasDatabaseUserByProjectsIndex
+}
+
+func (a *AtlasDatabaseUserByProjectsIndexer) Keys(object client.Object) []string {
+	user, ok := object.(*akov2.AtlasDatabaseUser)
+	if !ok {
+		a.logger.Errorf("expected *akov2.AtlasDatabaseUser but got %T", object)
+		return nil
+	}
+
+	if user.Spec.Project != nil && user.Spec.Project.Name != "" {
+		return []string{user.Spec.Project.GetObject(user.Namespace).String()}
+	}
+
+	if user.Spec.AtlasRef != nil && user.Spec.AtlasRef.ID != "" {
+		return []string{user.Spec.AtlasRef.ID}
```

Comment:
As mentioned in a previous review, this is is dangerous. We are conflating two types of keys in one index:
1. The name/namespace of a AKO `AtlasProject` pointing to AKO `AtlasDatabaseUser` instances.
2. Atlas IDs pointing to AKO `AtlasDatabaseUser` instances.

Reply:
(as mentioned previously I recommend two separate indexes to not conflate key types and values)

Reply:
xref https://github.com/mongodb/mongodb-atlas-kubernetes/pull/1769#discussion_r1721233559

Reply:
For the CEL tests, I'd rather do that in a separate PR. This PR is big enough. In fact, we could probably merge the CEL test infra already, as it is not public facing, and just reuse here.

Reply:
Agreed.